### PR TITLE
The Akka refactor we've all been waiting for :-)

### PR DESCRIPTION
**Pull request by fommil**
This refactor has been a long time coming. I'm reaping all the glory, but it has been enabled almost entirely by the extensive testing suites.

A few things I want to do before this is merge-worthy:

1. ~~use it in anger for a few days. This is a big change and regression risk is high, even with our tests.~~
2. ~~make sure that all the incoming SWANK messages have coverage (i.e. uncomment the lines in `SwankFormatsSpec`).~~
3. ~~JSON is fairly easy to add it now as a runtime alternative to SWANK.~~
4. ~~instead of failing immediately when the presentation compiler is busy, I'll always try to answer questions (maybe put a staleness checker in at some point)~~

### CODE REVIEW

**Comment by fommil** [comment]
OK, by the shear fact that this was assigned #1000 it means I **have** to get JSON working. It's the law, or something.

**Comment by rorygraves** [comment]
Build fails - pr rejected ;)

**Comment by fommil** [comment]
AppVeyor won't spam us again. Shippable timed out... I suppose we could add a time dilation factor.

File: core/src/main/scala/org/ensime/core/Refactoring.scala [hunk 1]
```diff
@@ -64,25 +62,22 @@ trait RefactoringHandler { self: Analyzer =>
...
-        project ! AddUndo(
-          "Refactoring of type: " + req.tpe.symbol.toString,
-          FileUtils.inverseEdits(effect.changes, charset)
-        )
         val result = scalaCompiler.askExecRefactor(procedureId, req.tpe, effect)
         sender ! result
       case None =>
         val f = RefactorFailure(procedureId, "No effect found for procId " + procedureId)
-        sender ! Left(f)
+        sender ! f
```
Comment (rorygraves) [inline core/src/main/scala/org/ensime/core/Refactoring.scala]: Might was well just do sender ! Refactor... here val f is not really needed

File: swank/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala [hunk 2]
```diff
@@ -487,6 +480,48 @@ object SwankProtocolResponse {
       case _ => deserializationError(hint)
     }
   }
+
+  // WORKAROUND not having a sealed family for RPC responses
+  def unhappyFamily(msg: Any): Sexp = msg match {
```
Comment (rorygraves) [inline swank/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala]: Lol

File: swank/src/test/scala/org/ensime/server/protocol/swank/SwankProtocolSpec.scala [hunk 3]
```diff
@@ -1,700 +1,81 @@
...
-    }
-
-    it("should understand swank:patch-source") {
-      testWithResponse("""(swank:patch-source "Analyzer.scala" (("+" 6461 "Inc") ("-" 7127 7128) ("*" 7200 7300 "Bob")))""") { (t, m, id) =>
-        (t.patchSource _).expects(file("Analyzer.scala").canon, List(PatchInsert(6461, "Inc"), PatchDelete(7127, 7128), PatchReplace(7200, 7300, "Bob")))
-        (m.send _).expects(s"""(:return (:ok t) $id)""")
-      }
-    }
-
-    it("should understand swank:typecheck-all") {
```
Comment (rorygraves) [inline swank/src/test/scala/org/ensime/server/protocol/swank/SwankProtocolSpec.scala]: Are their new equivalents of these tests?

**Comment by rorygraves** [comment]
Done a quick look through - will do a deeper dive over the next few days.
A bit of a brain dump of random comments
* I think its bad we lose the blocking api part - thought we were moving that rather than just removing it.
* I like the comments showing the message responses - could we capture that through types  - Foo extends RPCRequest[ResponseType]
* it feels like we have lost a bunch of swank protocol decoding tests - I think its important we have string type validation tests to ensure we don't break the encoding. 
We should also try and get coverage on (even if coveralls doesn't work to validate we haven't dropped testing.

**Comment by rorygraves** [comment]
What I failed to say is that it looks really good.

**Comment by fommil** [comment]
I'd want the blocking API to come back for the programmatic access but since I don't have any tests for it at the moment, I was concerned about just leaving it to rot.

**Comment by fommil** [comment]
damn, in emacs

```
ssertion failed at 'Test find-references.': (ensime-test-assert-failed . Expected truth of (search-forward class B(value:String) extends A nil t).)
Error executing test: (ensime-test-interrupted . Test interrupted: (ensime-test-assert-failed . Expected truth of (search-forward class B(value:String) extends A nil t).).), moving to next.
```

**Comment by fommil** [comment]
ok, I've done everything I set out to do with this PR. When it goes green I'm happy to recommend a merge :smile: (now lets hurry the hell up so we can release @ediani's awesome implicit stuff! :stuck_out_tongue: )

**Comment by rorygraves** [comment]
@aemoncannon @edani No objections from me - unless I hear one from you (or you merge it) I will merge in the morning (if the in-progress build turns green again)

**Comment by fommil** [comment]
if it doesn't turn green, I'll come round to your desk with a green felt tip pen and **MAKE IT GREEN**

**Comment by rorygraves** [comment]
Careful, I have my sword... and somebody did shapeless magic in my build...

**Comment by aemoncannon** [comment]
awesome work. sorry I'm a bit out of the loop: are elisp tests running at the moment?

**Comment by fommil** [comment]
pffh, build results by the morning? you're hopeful. This includes an upgrade to shapeless so see you sometimes in 2016.

**Comment by fommil** [comment]
@aemoncannon yup, they still run! Caught a few regressions earlier actually.

**Comment by rorygraves** [comment]
Btw, saw @edani Implicit stuff this morning - very very cool

**Comment by fommil** [comment]
actually, that reminds me, I need to rebase eric's stuff on top of my last commits in this PR.

**Comment by fommil** [comment]
@aemoncannon btw, we're still waiting on ScalaDays publishing the talk video. It was a really great even, wish you all had been there! Loads of interest in ENSIME.

**Comment by fommil** [comment]
GREEN!

**Comment by aemoncannon** [comment]
Can't wait to see it! Ensime is really growing up lately...

**Pull request merged**